docs/user_configuration.md
Django have built-in support for managing access to pages and makes it possible to have different access levels of users, please read more are [django's homepage](https://docs.djangoproject.com/en/3.2/topics/auth/default/)

We recommend that you set up at least two groups, admin and flowcell. Where admin can: 

- change/add users and flowcells
- delete uploaded flowcells.

| Name | Permission | Description |
| :--- | :--- | : --- |
| admin | <ul><li>auth &#124; user &#124; Can add user</li><li>auth &#124; user &#124; Can change user</li><li>auth &#124; user &#124; Can view user</li></ul> | Users belonging to this group have the ability <br /> to add, change and view users |
| flowcell | <ul><li>dataprocessor &#124; flowcell &#124; Can delete flowcell</li><li> dataprocessor &#124; flowcell &#124; Can view flowcell </li><li> dataprocessor &#124; samples run data &#124; Can<br />delete samples run data </li><li> dataprocessor &#124; batch run &#124; Can delete <br /> batch run </li></ul> | Users belonging to this group have the<br />ability to delete flowcells and the<br />information bound to the flowcell. |

** Note ** that the super user, created by running `python3 manage.py createsuperuser`, have access to everything and can do whatever he/she wants.

docs/server_configuration.md
NIPTviewer has been developed using django and most changes to the server can be made by modifying [niptviewer/settings.py](https://github.com/clinical-genomics-uppsala/NIPTviewer/blob/dev/niptviewer/niptviewer/settings.py), for more information please visist [django homepage](https://docs.djangoproject.com/en/3.2/ref/settings/).

Some of NIPTviewer settings that can and **should** be modified before running NIPTviewer in any kind of production environment. This can be accomplished by simply setting environment variables.

# ENV config

## SECRET_KEY
`SECRET_KEY` key is used to provide cryptographic signing. This key is mostly used to sign session cookies.

## Time selection
Default time span can be set for the different parts of NIPTviewer

### Web page
`DEFAULT_TIME_SELECTION` number of month from today that should be shown, default 9999 (show all data)

### Sample Report
`DEFAULT_TIME_SELECTION_SAMPLE_REPORT` number of month from today that should be shown in pdf report, default 24

### QC Report
`DEFAULT_TIME_SELECTION_QC_REPORT` number of month from today that should be shown in pdf report, default 12

## Database 

The default setting file currenlty support 3 types of databases, sqlite3, postgres and mssql

Settable variables are:

- DATABASE
- ENGINE
- NAME
- USER
- PASSWORD
- HOST
- PORT

the database section from the setting file
```python
if os.environ.get('DATABASE', "sqlite3") == "sqlite3":
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': BASE_DIR / 'db.sqlite3',
        }
    }
elif os.environ.get('DATABASE', "sqlite3") == "postgres":
    DATABASES = {
        "default": {
            "ENGINE": os.environ.get("SQL_ENGINE", "django.db.backends.sqlite3"),
            "NAME": os.environ.get("SQL_DATABASE", os.path.join(BASE_DIR, "db.sqlite3")),
            "USER": os.environ.get("SQL_USER", "user"),
            "PASSWORD": os.environ.get("SQL_PASSWORD", "password"),
            "HOST": os.environ.get("SQL_HOST", "localhost"),
            "PORT": os.environ.get("SQL_PORT", "5432"),
        }
    }
elif os.environ.get('DATABASE', "sqlite3") == "mssql":
    DATABASES = {
        "default": {
            "ENGINE": 'mssql',
            "NAME": get_secrets('SQL_DATABASE', os.path.join(BASE_DIR, 'db.sqlite3')),
            "USER": get_secrets('SQL_USER', 'user'),
            "PASSWORD": get_secrets('SQL_PASSWORD', 'password'),
            "HOST": os.environ.get('SQL_HOST', 'localhost'),
            "PORT": os.environ.get('SQL_PORT', ''),
            "OPTIONS": {
                'driver': 'ODBC Driver 17 for SQL Server',
                'host_is_server': True,
                'connection_timeout': 30,
                'collation': 'SQL_Latin1_General_CP1_CI_AS',
                'extra_params': 'TrustServerCertificate=yes;Encrypt=yes',
            }
        }
    }
```

Support for more database can easily be added by modifying this section of settings.py



CHANGELOG.md
# Changelog

## [1.10.0](https://github.com/clinical-genomics-uppsala/NIPTviewer/compare/v1.9.3...v1.10.0) (2023-12-15)


### Features

* docs setup ([547071c](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/547071c178d1788ab6b085b517e72501765558ec))
* gitpage setup ([18054a3](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/18054a39672f6364ccb6e8c2984a1f6e583c8bb1))
* more plots and pdf files ([c054402](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/c05440292745cdbd48640177d01053943ca74860))
* second commit ([3b6fe33](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/3b6fe338ea6714b984a3e79cfa0a1cac3a19de28))
* working plots ([807c9f9](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/807c9f9ee02b0e7388f114682013c47ea20b81dd))


### Bug Fixes

* add missing value ([98e5fb2](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/98e5fb26be75fc56a752dbab8a593426a3edba30))
* minor bug fixes handling corner cases ([c9985e3](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/c9985e3ff12f0e4f0f2e706bc38e19b49867384b))
* update docker env ([6dd5c69](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/6dd5c6945708a7c84dd3fe5cd5b6e91153436da5))
* update git version ([aa38d7a](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/aa38d7a78b34f8cd40206ce4aac5be89fba9e660))
* update version for curl, python3-pip and microsoft sql ([6bdb0ba](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/6bdb0ba44d58e9004f0948ddd7d025d3e41948fc))


### Documentation

* update name format in mkdocs ([15430c9](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/15430c94c536556dd4af00744ab78e0a627d6660))
* update readme ([c8fbed6](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/c8fbed6eb3ba5513faa1fee9f2e94b52c78001d1))
* update readthedocs ([bf4f855](https://github.com/clinical-genomics-uppsala/NIPTviewer/commit/bf4f85529504b3064389b922d8ecba38eedea5bb))

## [1.9.3](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.9.2...v1.9.3) (2022-12-01)


### Bug Fixes

* add missing value ([694685f](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/694685f9be9d40879897dc7770f827afe9d7bccf))
* update gnupg2 dependency in dockerfile ([22596db](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/22596dbfcc8e3879ef06aa8ef11c60640dba34db))
* update libpq-dev and vim dependency in dockerfile ([860a086](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/860a086d79a70a54221085f7733f45b240220223))

### [1.9.2](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.9.1...v1.9.2) (2022-05-23)


### Bug Fixes

* add missing data to report and make sample page load ([ae50904](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/ae50904fdaea4203eb9e1f773f009f7719788af4))

### [1.9.1](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.9.0...v1.9.1) (2022-05-16)


### Bug Fixes

* add missing % sign ([d8c5b9b](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/d8c5b9b43ebb87b4c653bcf11f5d37a91db69794))
* minor modifications for test env. ([64a6edc](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/64a6edc22b1490828ccf9e7ef01a1b94dd5f594c))
* revert curl removal ([b42c509](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/b42c509df39f80ca98e9df1ac1bee73eea7cb2d9))
* skip curl, prevents container from building ([75fd50e](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/75fd50e70ea65c4b2f5a6a643b3943e32f4c6e68))


### Performance Improvements

* update code to improve performance during report generation ([326ba9a](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/326ba9a4f073836c777e2e2d2ae64ef1df2ce371))

## [1.9.0](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.8.2...v1.9.0) (2022-05-06)


### Features

* add test kube files and update nginx image ([fd85205](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/fd852051a044594c378c6edb601199868c35bb6f))


### Bug Fixes

* update Dockerfile to handle missing version of curl ([cfb048d](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/cfb048de44abbb6f9c29d8a4f676f1d82bde688a))
* update nginx ([f192433](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/f1924336d7b1c8e84a9e006b866aa0c01942fc47))

### [1.8.2](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.8.1...v1.8.2) (2022-03-24)


### Bug Fixes

* handle migration between different database better. ([7bd0cdc](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/7bd0cdc6be525bdb2d5f2032fe5d4a8ff67c0428))

### [1.8.1](https://github.com/clinical-genomics-uppsala/NIPTViewer/compare/v1.8.0...v1.8.1) (2022-03-23)


### Bug Fixes

* add missing admin templates. ([#142](https://github.com/clinical-genomics-uppsala/NIPTViewer/issues/142)) ([a2d5288](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/a2d52889ac3e2c16d0b66163168a4e41d60d76a9))
* add missing admin templates. ([#143](https://github.com/clinical-genomics-uppsala/NIPTViewer/issues/143)) ([c5d1b96](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/c5d1b96583e562c5ebe2293f502ead2f3f4b56d9))
* turn of debug on produktion files ([1e566c8](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/1e566c80583ba3ffdd8054b8a2dc2c33e8ac18c5))
* update readme ([70b9ed0](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/70b9ed0d9f5cc3b297f88500adeb61df0c0f2115))
* update readme ([#144](https://github.com/clinical-genomics-uppsala/NIPTViewer/issues/144)) ([55d1a8d](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/55d1a8d5325882b308619e01f92f3769303d513d))
* update readme ([#147](https://github.com/clinical-genomics-uppsala/NIPTViewer/issues/147)) ([417a28a](https://github.com/clinical-genomics-uppsala/NIPTViewer/commit/417a28af9dc5c1e0809adde24dd2d6e2c93bf22d))


niptviewer/niptviewer/settings.py
"""
Django settings for niptviewer project.

Generated by 'django-admin startproject' using Django 3.1.1.

For more information on this file, see
https://docs.djangoproject.com/en/3.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.1/ref/settings/
"""

from pathlib import Path
import os
import logging.config

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.1/howto/deployment/checklist/

DEBUG = int(os.environ.get("DEBUG", default=0))

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get("SECRET_KEY", None if DEBUG == 0 else "DUMMY")

# SECURITY WARNING: don't run with debug turned on in production!

DEFAULT_TIME_SELECTION = os.environ.get("TIME_SELECTION", 9999)
DEFAULT_TIME_SELECTION_SAMPLE_REPORT = os.environ.get("SELECTION_SAMPLE_REPORT", 24)
DEFAULT_TIME_SELECTION_QC_REPORT = os.environ.get("TIME_SELECTION_QC_REPORT", 12)

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'reportvisualiser',
    'dataprocessor',
    'users',
    'wkhtmltopdf'
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'niptviewer.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates')],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'niptviewer.context_processors.niptviewer_version',
            ],
        },
    },
]


def get_secrets(variable, default):
    if os.path.isfile(os.getenv(variable, default)):
        data = ""
        with open(os.getenv(variable), "r") as myfile:
            data = myfile.readlines()
        return data
    else:
        return os.getenv(variable, default)


# Database
# https://docs.djangoproject.com/en/3.1/ref/settings/#databases

if os.environ.get('DATABASE', "sqlite3") == "sqlite3":
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': BASE_DIR / 'db.sqlite3',
        }
    }
elif os.environ.get('DATABASE', "sqlite3") == "postgres":
    DATABASES = {
        "default": {
            "ENGINE": os.environ.get("SQL_ENGINE", "django.db.backends.postgresql"),
            "NAME": os.environ.get("SQL_DATABASE", os.path.join(BASE_DIR, "db.sqlite3")),
            "USER": os.environ.get("SQL_USER", "user"),
            "PASSWORD": os.environ.get("SQL_PASSWORD", "password"),
            "HOST": os.environ.get("SQL_HOST", "localhost"),
            "PORT": os.environ.get("SQL_PORT", "5432"),
        }
    }
elif os.environ.get('DATABASE', "sqlite3") == "mssql":
    DATABASES = {
        "default": {
            "ENGINE": 'mssql',
            "NAME": get_secrets('SQL_DATABASE', os.path.join(BASE_DIR, 'db.sqlite3')),
            "USER": get_secrets('SQL_USER', 'user'),
            "PASSWORD": get_secrets('SQL_PASSWORD', 'password'),
            "HOST": os.environ.get('SQL_HOST', 'localhost'),
            "PORT": os.environ.get('SQL_PORT', ''),
            "OPTIONS": {
                'driver': 'ODBC Driver 17 for SQL Server',
                'host_is_server': True,
                'connection_timeout': 30,
                'collation': 'SQL_Latin1_General_CP1_CI_AS',
                'extra_params': 'TrustServerCertificate=yes;Encrypt=yes',
            }
        }
    }

WSGI_APPLICATION = 'niptviewer.wsgi.application'

# Password validation
# https://docs.djangoproject.com/en/3.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/3.1/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.1/howto/static-files/

STATIC_URL = "/staticfiles/"
STATIC_ROOT = os.path.join(BASE_DIR, "staticfiles")
STATICFILES_DIRS = (
  'assets/',
)

STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

WKHTMLTOPDF_CMD_OPTIONS = {
    'quiet': True,
}

LOGIN_REDIRECT_URL = "/"
LOGIN_URL = '/login'
LOGOUT_URL = '/logout'
LOGOUT_REDIRECT_URL = '/login'

LimitRequestBody: 102400

CSRF_COOKIE_SECURE: True

SESSION_COOKIE_SECURE: True

ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS", "127.0.0.1").split(" ")

DEFAULT_FROM_EMAIL = os.environ.get('FROM_EMAIL', "DUMMAY")

SERVER_EMAIL = os.environ.get('SERVER_EMAIL', "DUMMY")

LOGGING_CONFIG = None
#
# # Get loglevel from env
LOGLEVEL = os.getenv('DJANGO_LOGLEVEL', 'debug').upper()
